docs/plugin_developers_guide/02_writing_plugins.rst

```diff
@@ -242,3 +242,111 @@ In C/C++ the *plugin_reconfigure* class is very similar, note however that the *
+
+  - A descriptive way to define the configuration such that user interfaces can be built without prior knowledge of the elements to be configured.
+
+  - A common approach that will work across many different languages.
+
+Fledge divides its configuration in categories. A category being a collection of configuration items. A category is also the smallest item of configuration that can be subscribed to by the code. This subscription mechanism is they way that Fledge facilitates dynamic reconfiguration. It allows a service to subscribe to one or more configuration categories, whenever an item within a category changes the central configuration manager will call a handler to pass the newly updated configuration category. This handler my be within a services or between services using the micro service management API that every service must support. The mechanism however is transparent to the code involved.
+
+The configuration items within a category are JSON object, the object key is the name of the configuration item, the object itself contains data about that item. As an example, if we wanted to have a configuration item called *MaxRetries* that is an integer with a default value of 5, then we would configured it using the JSON object
+
+.. code-block:: JSON
+
+   "MaxRetries" : {
+                "type" : "integer",
+                "default" : "5"
+                }
+
+We have used the properties *type* and *default* to define properties of the configuration item *MaxRetries*.  These are not the only properties that a configuration item can have, the full set of properties are
+
+.. list-table::
+   :header-rows: 1
+
+   * - Property
+     - Description
+   * - type
+     - The type of the configuration item. The list of types supported are; integer, float, string, enumeration, boolean, JSON, URL, script, code and northTask.
+   * - default
+     - The default value for the configuration item. This is always expressed as a string regardless of the type of the configuration item.
+   * - value
+     - The current value of the configuration item. This is not included when defining a set of default configuration in, for example, a plugin.
+   * - displayName
+     - The string to use in the user interface when presenting the configuration item. Generally a more user friendly form of the item name. Item names are referenced within the code.
+   * - description
+     - A description of the configuration item used in the user interface to give more details of the item. Commonly used as a mouse over help prompt.
+   * - order
+     - Used in the user interface to give an indication of how high up in the dialogue to place this item.
+   * - options
+     - Only used for enumeration type elements. This is a JSON array of string that contains the options in the enumeration.
+   * - validity
+     - An expression used to determine if the configuration item is valid. Used in the UI to gray out one value based on the value of others.
+
```

Comment:
We have also supported optional attributes like `readonly, length, maximum, minimum, rule, deprecated, mandatory`. Better to document all these.